src/handlers/actions/managePermissions.ts

```diff
@@ -0,0 +1,127 @@
+import { mutate, query } from '~amplifyClient';
+import { ContractEvent, ColonyActionType } from '~types';
+import {
+  getColonyRolesDatabaseId,
+  getDomainDatabaseId,
+  createInitialColonyRolesDatabaseEntry,
+  createColonyHistoricRoleDatabaseEntry,
+  getAllRoleEventsFromTransaction,
+  getRolesMapFromEvents,
+  verbose,
+  writeActionFromEvent,
+} from '~utils';
+
+export default async (event: ContractEvent): Promise<void> => {
+  const { args, contractAddress, blockNumber, transactionHash } = event;
+  const { agent, user: targetAddress, domainId } = args;
+
+  const id = getColonyRolesDatabaseId(contractAddress, domainId.toString(), targetAddress);
+  const domainDatabaseId = getDomainDatabaseId(contractAddress, domainId.toString());
+
+  /*
+   * @TODO
+   * - create action list entry
+   */
+
+  const {
+    id: existingColonyRoleId,
+    latestBlock: existingColonyRoleLatestBlock,
+    ...existingRoles
+  } = (await query('getColonyRole', { id })) || {};
+
+  /*
+   * update the entry
+   */
+  if (existingColonyRoleId) {
+    /*
+     * Only update the entry if the event we're processing is older than the latest
+     * permissions we stored in the database
+     *
+     * Would be ideal if we could tell if the previous mutation executed, because as it stands,
+     * until it finishes, this logic check can't really tell and will try to execute it again
+     *
+     * (it doesn't break anything, as the GraphQL server will just discard it with an error,
+     * but it just adds more travel time which is wasted)
+     */
+    if (blockNumber > parseInt(existingColonyRoleLatestBlock, 10)) {
+      const allRoleEventsUpdates = await getAllRoleEventsFromTransaction(transactionHash, contractAddress);
```

Comment:
Maybe leave a comment here as this deviates from the general pattern of having one handler per event. I was scratching my head as to why we even need to do this, before realising it's because we have one handler here for multiple events.

Reply:
If it sounds like I'm off or would be easier to explain on a call, I'd happy  jump on one to get a better understanding of how this all fits together (and would probably be beneficial for everyone also)